docs/weave/design/03-permissions.md

```diff
@@ -0,0 +1,156 @@
+
+## Authentication
+
+Authentication information is added to the context as part of the middleware stack, and used to verify the caller. The simplest example is signature verification. We check if the signature validates against a known public key, and after checking nonces for replay protection, can authenticate this public key for this transaction.
+
+However, Ethereum devs are used to the concept of authority not just being tied to a signature, but potentially a smart contract. We will allow something similar, but we don't need to be as general, as we also don't have the same general "anyone can call anything" architecture, nor do we run untrusted code.
+
+We use multiple middlewares to check for various conditions on the transaction and add the authentication information to the `Context`. The basic example, `x/sigs.Middleware`, checks if the Tx has signatures, and if so validates them.
+
+```go
+// SignedTx represents a transaction that contains signatures,
+// which can be verified by the auth.Decorator
+type SignedTx interface {
+    // GetSignBytes returns the canonical byte representation of the Msg.
+    // Equivalent to weave.MustMarshal(tx.GetMsg()) if Msg has a deterministic
+    // serialization.
+    //
+    // Helpful to store original, unparsed bytes here, just in case.
+    GetSignBytes() ([]byte, error)
+
+    // Signatures returns the signature of signers who signed the Msg.
+    GetSignatures() []*StdSignature
+}
+```
+
+It stores the matching conditions in the context under a secret key, and exposes an `Authenticator` that can be used to read this information.
+
+```go
+type contextKey int // local to the auth module
+
+const (
+    contextKeySigners contextKey = iota
+)
+
+// withSigners is a private method, as only this module
+// can add a signer
+func withSigners(ctx weave.Context, signers []weave.Condition) weave.Context {
+    return context.WithValue(ctx, contextKeySigners, signers)
+}
+
+// Authenticate implements x.Authenticator and provides
+// authentication based on public-key signatures.
+type Authenticate struct{}
+
+var _ x.Authenticator = Authenticate{}
+
+// GetConditions returns who signed the current Context.
+// May be empty
+func (a Authenticate) GetConditions(ctx weave.Context) []weave.Condition {
+    // (val, ok) form to return nil instead of panic if unset
+    val, _ := ctx.Value(contextKeySigners).([]weave.Condition)
+    // if we were paranoid about our own code, we would deep-copy
+    // the signers here
+    return val
+}
+```
+
+And finally, when we create a module that needs to read authentication info, we can pass in the handler, so it can use check authentication info from
```

Comment:
```suggestion
And finally, when we create a module that needs to read authentication info, we can pass in the handler, so it can use check authentication info from this middleware.
```